### ISSUE TITLE
Bug: adding properties to an object after the config is immutable shouldn't be possible.

### ISSUE THREAD

**Issue by Overdrivr** [node-config/node-config#471]
After investigating a nasty bug in my application, I realized that modifying a local instance of a config property also modifies the global config object, meaning that this property will be modified for further calls to `get`.

If this is a feature (IMO it should not), it should be either documented, or disabled by default.
Otherwise, it means the `config` cannot be really trusted as a single source of truth for configuration.
  
### I'm submitting a ... 
  - [x ] bug report
  - [  ] feature request

### What is the current behavior?
Repro

> ./config/default.json
```
{
  "testprop":{ }
}
```

```
const config = require('config');
a = config.get('testprop');
console.log(a);
a.delay = 123;
console.log(a);
console.log(config.get('testprop'));
```

Ouput
```
{}
{ delay: 123 }
{ delay: 123 }
```

### What is the expected behavior?
The object returned by `get` should be a deep copy of the global configuration. Modifying it should have no repercussions, and calls to `get` will always consistently return the same values. 

Expected output

```
{}
{ delay: 123 }
{}
```

### Please tell us about your environment:

 -  node-config version: config@1.29.4
 -  node-version: 8.9.4

**Comment by markstos** [comment]
`node-config` strives to be immutable after the first call to `config.get()`. I believe this case happens because it's not a case of mutating an existing property but adding a new one.

@Overdrivr Do you want to review the code to see if there's a place where we should call [Object.preventExtensions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/preventExtensions) as part of making the object immutable?

**Comment by Overdrivr** [comment]
Thanks for clearing that up !

Well I see two options, either preventingExtensions on the config/module object, in that case the first call to get seems appropriate for doing that (https://github.com/lorenwest/node-config/blob/master/lib/config.js#L200). I'm not sure what side effects there could be though.

Or the other option would be to freeze the value object returned by get, in that case it would rather be line 208 https://github.com/lorenwest/node-config/blob/master/lib/config.js#L208

I'm not sure which one is best, any ideas ?

**Comment by markstos** [comment]
What about adding it to `makeImmutable`?
https://github.com/lorenwest/node-config/blob/master/lib/config.js#L530

**Comment by Overdrivr** [comment]
Makes even more sense :) I'll make a PR today

**Comment by markstos** [comment]
> After investigating a nasty bug in my application, I realized that modifying a local instance of a config property also modifies the global config object

I agree we want to make config object immutable as soon as possible. Perhaps in the future this will be addressed by an overhaul like #569

But I'm closing this as stale. A related PR was submitted, but no one fixed the tests or conflicts in more than 4 years.
